extensions/route_map/lib.rs

```diff
@@ -0,0 +1,358 @@
+    fn add_routes<'a>(&mut self, py: Python<'a>, routes: Vec<wrappers::Route>) -> PyResult<()> {
+        for route in routes {
+            self.add_route(py, route)?;
+        }
+        Ok(())
+    }
+
+    /// Add a route to the map
+    #[pyo3(text_signature = "($self, route)")]
+    fn add_route<'a>(&mut self, py: Python<'a>, route: wrappers::Route<'a>) -> PyResult<()> {
+        let path = route.path()?;
+        let path_parameters: &PyAny = route.path_parameters()?;
+        let is_static = self.static_paths.contains(path);
+
+        let new_handler_group = if is_static {
+            if !route.is_asgi(&self.ctx)? {
+                return Err(wrappers::ImproperlyConfiguredException::new_err(format!(
+                    "Static route handlers must be asgi handlers at {path}"
+                )));
+            }
+            if path_parameters.is_true()? {
+                return Err(wrappers::ImproperlyConfiguredException::new_err(format!(
+                    "Static routes may not have path parameters at {path}"
+                )));
+            }
+            HandlerGroup::Static {
+                path: String::from(path),
+                handler: self
+                    .ctx
+                    .build_middleware_stack(py, route, route.handler()?)?,
+            }
+        } else if route.is_http(&self.ctx)? {
+            let mut asgi_handlers = HashMap::new();
+            for item in route.http_handlers()? {
+                let (method, handler) = item?;
+                asgi_handlers.insert(
+                    HandlerType::from_http_method(method),
+                    self.ctx.build_middleware_stack(py, route, handler)?,
+                );
+            }
+            HandlerGroup::NonAsgi {
+                path_parameters: path_parameters.into(),
+                asgi_handlers,
+            }
+        } else if route.is_asgi(&self.ctx)? {
+            HandlerGroup::Asgi {
+                path_parameters: path_parameters.into(),
+                handler: self
+                    .ctx
+                    .build_middleware_stack(py, route, route.handler()?)?,
+            }
+        } else if route.is_websocket(&self.ctx)? {
+            HandlerGroup::websocket(
+                path_parameters.into(),
+                self.ctx
+                    .build_middleware_stack(py, route, route.handler()?)?,
+            )
+        } else {
+            return Err(PyTypeError::new_err(format!(
+                "Unknown route type {}",
+                route.type_name()?,
+            )));
+        };
+
+        search::find_insert_handler_group(
+            &mut self.root,
+            &mut self.plain_routes,
+            path,
+            path_parameters,
+            is_static,
+            new_handler_group,
+        )?;
+
+        Ok(())
+    }
+
+    /// Given a scope, retrieves the correct ASGI App for the route
+    fn resolve_asgi_app(&self, py: Python<'_>, scope: wrappers::Scope) -> PyResult<Py<PyAny>> {
```

Comment:
the `<'_>` here can be elided as well

Reply:
I go back and forth on this. There is a [lint](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#elided-lifetimes-in-paths) for this as well. I can remove it though.